scripts-dev/complement.sh

```diff
@@ -31,11 +32,12 @@ cd "$(dirname $0)/.."
 
 # Check for a user-specified Complement checkout
 if [[ -z "$COMPLEMENT_DIR" ]]; then
```

Comment:
I'm curious why using `COMPLEMENT_DIR` won't work for you? Do you not want to pull the entire repo for some reason? (It is relatively small IIRC.)

Reply:
Well, it is easier to update a single environment variable than having to add "install git, clone repo, checkout right commit, set environment variable" to our CI. It's not like we can't do that, but having to enabled/disable that every time the upstream complement doesn't work with the current synapse release is more effort than just telling it to use the right commit. I also just figured out how to do that correctly, after I hacked in the support for specifying a ref and using a ref is less effort. It might also be useful in the future, if complement ever gets branches for synapse versions like "synapse-1.50" or so (because sometimes the branches are just incompatible).

Reply:
I can also easily run the complement job in our CI with the complement ref overwritten, which makes it easier to figure out what commit actually works. We had to do that today, because the complement CA changes somehow don't work in our CI and we couldn't figure out why yet.

Reply:
> Well, it is easier to update a single environment variable than having to add "install git, clone repo, checkout right commit, set environment variable" to our CI.

Boiled down, this PR is the equivalent of adding the listed steps inside of `complement.sh`. One could also just easily add a "install git, clone repo, checkout right commit, set environment variable" bit to their dockerfile, and control which commit is checked out with another environment variable. And yes, that would work without adding another env var to `complement.sh`.

But this PR proposes upstreaming it as optional functionality into the script, so that every project doesn't need to set that up their own. It's a convenience that could save downstream developers time, so while it's not a necessary option (and is code we'll need to maintain), I think it's justifiable.

Reply:
It's also a really small change, because it just replaces the hardcoded master with an overrideable variable.

Reply:
> It might also be useful in the future, if complement ever gets branches for synapse versions like "synapse-1.50" or so (because sometimes the branches are just incompatible).

We do branch matching in our CI for this reason, see: https://github.com/matrix-org/synapse/blob/c072c0b82969879cd08d0d6e669ded509f1e833b/.github/workflows/tests.yml#L342-L364

We don't use `complement.sh` directly for Synapse CI, I'm not sure it is really designed for such things. Anyway, I agree that the overall change is relatively minor.

Reply:
Hm, I guess we should do that too at some point then. I was mostly happy we got it working at all :D